### DISCUSSION TITLE
JAX fashion of interleave concatenation

### DISCUSSION THREAD

**Discussion by zgbkdlm** [jax-ml/jax#13668] [Q&A] [answered]
Let `a` and `b` be two arrays

```
a = jnp.array([1, 3, 5, 7])
b = jnp.array([0, 2, 4, 6])
```

We aim to create an array `c = jnp.array([0, 1, 2, 3, 4, 5, 6, 7])` by interleave-concatenating `a` and `b`. In Numpy it's simple like

```
c = np.zeros((8,))
c[0::2] = b
c[1::2] = a
```

What is an efficient JAX way for this concatenation?

**Comment by soraros** [comment, marked as answer]
There are many ways to do this, the following seems to be the simplest:

```python
@jit
def interleave_concat(a, b):
  return jnp.column_stack([a, b]).reshape(-1)
```

One could make use of the [functional equivalence](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html?highlight=at) of the numpy in-place array modification operations as follows:

```python
@jit
def interleave_concat(a, b):
  res = jnp.zeros(len(a) * 2, dtype=a.dtype)
  return res.at[0::2].set(a).at[1::2].set(b)
```

**Reply by zgbkdlm** [reply]
Is there any trade-off using the in-place updates? I recall that the in-place updates were not well supported.

**Reply by jakevdp** [reply]
Regarding in-place updates: the short version is that outside JIT, the updates will lead to array copies (because JAX arrays are immutable at the Python level). Within JIT, the compiler will apply updates in-place when possible (e.g. if the original array is not used again). A bit more info here: https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#array-updates-x-at-idx-set-y

**Reply by zgbkdlm** [reply]
> There are many ways to do this, the following seems to be the simplest:
> 
> ```python
> @jit
> def interleave_concat(a, b):
>   return jnp.column_stack([a, b]).reshape(-1)
> ```


It doesn't work if the lengths of `a` and `b` are different. E.g., `a = jnp.array([1, 3, 5, 7])` b = jnp.array([0, 2, 4])`
